### PR TITLE
[BUG-1402] Return HTTP 409 on version confict when creating & updating tenants - added relevant unit test

### PR DESCRIPTION
**Pull request by prabhask5**
### Description
Added new TenantsActionApiTest file to put api unit tests for the /api/tenants api call- additionally wrote new unit test to test for parallel PUT /tenant/{name} calls, that they successfully return a 409 call and retry the call.

* Category (Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation)
Test fix

* Why these changes are required?
There was no test that checked for the behavior mentioned in the linked ticket, so previously we didn't know if the issue still existed or not (it apparently was fixed before), to make sure this doesn't happen again for this particular case, I decided to write a concrete test for it.

* What is the old behavior before changes and new behavior after changes?
Test was added, there's no behavior change?

### Issues Resolved
- #1402 - added unit test to cover bug in this issue

### Testing
Added new tenant action unit test to test parallel tenant api calls.

### Check List
- [x] New functionality includes testing
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prabhask5** [comment]
@peternied I'm getting the same error in these test where the CI tests cannot find package org.opensearch.test.framework, very confused on why this is happening since the local tests are working.

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java [hunk 1]
```diff
@@ -12,16 +12,27 @@
...
 import org.apache.hc.core5.http.Header;
-import org.apache.hc.core5.http.HttpStatus;
+import org.apache.http.HttpStatus;
 import org.junit.Assert;
 import org.junit.Test;
 
 import org.opensearch.common.settings.Settings;
 import org.opensearch.security.support.ConfigConstants;
 import org.opensearch.security.test.helper.rest.RestHelper;
+import org.opensearch.test.framework.AsyncActions;
```
Comment (peternied) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:22]: (!) Sorry I didn't see this sooner, this package is only available to tests that are in the `integrationTest` project.  You'll need to port the new test over to that framework.

File: src/integrationTest/java/org/opensearch/security/rest/TenantAsyncActionTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
+package org.opensearch.security.rest;
```
Comment (cwperks) [inline src/integrationTest/java/org/opensearch/security/rest/TenantAsyncActionTest.java:1]: Can you add a license header for this file?

```
/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 */
```

File: src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java [hunk 3]
```diff
@@ -12,7 +12,7 @@
 package org.opensearch.security.dlic.rest.api;
 
 import org.apache.hc.core5.http.Header;
-import org.apache.hc.core5.http.HttpStatus;
+import org.apache.http.HttpStatus;
```
Comment (cwperks) [inline src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java:15]: IMO imports in the `test` package should use `hc.core5`. The issue with automatic backporting is most pronounced in files under the `integrationTest` package.

File: src/integrationTest/java/org/opensearch/security/rest/TenantAsyncActionTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+        AtomicInteger numCreatedResponses = new AtomicInteger();
+        AsyncActions.getAll(conflictingRequests, 1, TimeUnit.SECONDS).forEach((response) -> {
+            assertThat(response.getStatusCode(), anyOf(equalTo(HttpStatus.SC_CREATED), equalTo(HttpStatus.SC_CONFLICT)));
+            if (response.getStatusCode() == HttpStatus.SC_CREATED) numCreatedResponses.getAndIncrement();
+        });
+        assertThat(numCreatedResponses.get(), equalTo(1)); // should only be one 201
+
+        RestHelper.HttpResponse getResponse = rh.executeGetRequest(TENANTS_ENDPOINT); // make sure the one 201 works
+        assertThat(getResponse.findValueInJson("tenant1" + ".description"), equalTo("create new tenant"));
```
Comment (cwperks) [inline src/integrationTest/java/org/opensearch/security/rest/TenantAsyncActionTest.java:59]: nit: The string concat looks unnecessary here.

**Review by cwperks — COMMENTED** [review]
Thank you @prabhask5. This PR looks good to me. I left a few comments to address and will approve once all comments have been addressed.

File: src/integrationTest/java/org/opensearch/security/rest/TenantAsyncActionTest.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    }
+}
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/rest/TenantAsyncActionTest.java:61]: nit: new line at the end of the file

**Review by DarshitChanpura — COMMENTED** [review]
There are some styling/import errors in CI. Could you please address those?
This PR looks good to me otherwise

**Comment by prabhask5** [comment]
@peternied I'm getting errors now that can't resolve `import org.opensearch.security.dlic.rest.api.AbstractRestApiUnitTest;`, is this a similar issue where I can't use this package outside of the rest/api folder?

**Comment by peternied** [comment]
@prabhask5 Please look into the test failures

**Comment by DarshitChanpura** [comment]
@prabhask5 Are you still working on this?

**Comment by peternied** [comment]
Doesn't look like this is being worked on anymore, going to close this out. Please feel free to reopen if you re-engage on this @prabhask5

**Pull request closed without merging**